File: TeeNew/Builder2006/Tool_SurfaceNearest.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#include <math.h>
#pragma hdrstop

#include "Tool_SurfaceNearest.h"
#include "TeeSurfaceTool.hpp"
#include "EditChar.hpp"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "Base"
#pragma link "TeeSurfa"
#pragma link "TeCanvas"
#pragma link "TeeTools"
#pragma resource "*.dfm"
TSurfaceNearestForm *SurfaceNearestForm;
//---------------------------------------------------------------------------
__fastcall TSurfaceNearestForm::TSurfaceNearestForm(TComponent* Owner)
  : TBaseForm(Owner)
{
}
//---------------------------------------------------------------------------
// This event is called when the user moves the mouse
// over a surface's cell.
// Let's use it in this demo to show the "hit" cell row and column.
void __fastcall TSurfaceNearestForm::ToolSelect(TObject *Sender)
{
  double tmpRow, tmpCol;
  int tmp = dynamic_cast<TSurfaceNearestTool*>(Sender)->SelectedCell;
  if (tmp ==-1) Label1->Caption = "";
  else
  {
    dynamic_cast<TSurfaceNearestTool*>(Sender)->GetRowCol(tmpRow,tmpCol);
    Label1->Caption = "Row: "+FloatToStr(tmpRow)+" Column: "+FloatToStr(tmpCol);
  }
}
//---------------------------------------------------------------------------
void __fastcall TSurfaceNearestForm::AddSurfacePoints()
{
  const int NumX = 50;
  const int NumZ = 30;

  Series1->NumXValues = NumX;
  Series1->NumZValues = NumZ;

  for (int x=1; x<=NumX; x++)
  {
    for (int z=1; z<=NumZ; z++)
    {
      Series1->AddXYZ( x,
                          cos(0.14*x)*cos(0.25*z)+
                          0.01*(x-(NumX/2))*(z-(NumZ/2)) +
                          13*exp(-0.06*(pow(x-(NumX/2),2)+pow(z-(NumZ/2),2)))+
                          6*exp(-0.03*(pow(x-(NumX/3),2)+pow(z-3*(NumZ/5),2)))-1,
                          z );
    }
  }
}
//---------------------------------------------------------------------------
void __fastcall TSurfaceNearestForm::FormCreate(TObject *Sender)
{

  Chart1->Axes->FastCalc = true;
  Chart1->View3DOptions->ZoomText = ztNo;

  AddSurfacePoints();

  // Create tool  (this can also be done at design-time without any code)
  TSurfaceNearestTool *tmp = new TSurfaceNearestTool(this);
  tmp->Series = Series1;
  tmp->OnSelectCell = ToolSelect;

  Chart1->Tools->Add(tmp);
}
//---------------------------------------------------------------------------

void __fastcall TSurfaceNearestForm::Button1Click(TObject *Sender)
{
  EditChartTool(this,Chart1->Tools->Items[0]);
}
//---------------------------------------------------------------------------

File: TeeNew/Tool_SurfaceNearest.pas
unit Tool_SurfaceNearest;
{$I TeeDefs.inc}

interface

uses
  {$IFNDEF LINUX}
  Windows, Messages,
  {$ENDIF}
  SysUtils, Classes,
  {$IFDEF CLX}
  QGraphics, QControls, QForms, QDialogs, QExtCtrls, QStdCtrls, QComCtrls,
  {$ELSE}
  Graphics, Controls, Forms, Dialogs, ExtCtrls, StdCtrls, ComCtrls,
  {$ENDIF}
  Base, TeeProcs, TeEngine, Chart, TeeSurfaceTool, TeeSurfa, TeeTools;

type
  TSurfaceNearestForm = class(TBaseForm)
    Button1: TButton;
    Series1: TSurfaceSeries;
    Label1: TLabel;
    ChartTool1: TRotateTool;
    procedure FormCreate(Sender: TObject);
    procedure Button1Click(Sender: TObject);
    procedure FormShow(Sender: TObject);
  private
    { Private declarations }
    Procedure AddSurfacePoints;
    procedure ToolSelect(Sender: TObject);
  public
    { Public declarations }
  end;

implementation

{$IFNDEF CLX}
{$R *.dfm}
{$ELSE}
{$R *.xfm}
{$ENDIF}

uses
  TeCanvas, EditChar;

procedure TSurfaceNearestForm.FormCreate(Sender: TObject);
var tmp : TSurfaceNearestTool;
begin
  inherited;

  Chart1.Axes.FastCalc:=True;
  Chart1.View3DOptions.ZoomText:=ztNo;

  AddSurfacePoints;

  // Create tool  (this can also be done at design-time without any code)
  tmp:=TSurfaceNearestTool.Create(Self);
  tmp.Series:=Series1;
  tmp.OnSelectCell:=ToolSelect;

  Chart1.Tools.Add(tmp);
end;

Procedure TSurfaceNearestForm.AddSurfacePoints;
const NumX=50;
      NumZ=30;
var x,z : Integer;
begin

  Series1.NumXValues:=NumX;
  Series1.NumZValues:=NumZ;

  for x:=1 to NumX do
      for z:=1 to NumZ do
          Series1.AddXYZ( x,
                          cos(0.14*x)*cos(0.25*z)+
                          0.01*(x-(NumX div 2))*(z-(NumZ div 2)) +
                          13*exp(-0.06*(sqr(x-(NumX div 2))+sqr(z-(NumZ div 2))))+
                          6*exp(-0.03*(sqr(x-(NumX div 3))+sqr(z-3*(NumZ div 5))))-1,
                          z );
end;

// This event is called when the user moves the mouse
// over a surface's cell.
// Let's use it in this demo to show the "hit" cell row and column.

procedure TSurfaceNearestForm.ToolSelect(Sender: TObject);
var tmp : Integer;
    tmpRow,
    tmpCol : Double;
begin
  tmp:=(Sender as TSurfaceNearestTool).SelectedCell;

  if tmp=-1 then Label1.Caption:=''
  else
  begin
    (Sender as TSurfaceNearestTool).GetRowCol(tmpRow,tmpCol);
    Label1.Caption:='Row: '+FloatToStr(tmpRow)+' Column: '+FloatToStr(tmpCol);
  end;
end;

procedure TSurfaceNearestForm.Button1Click(Sender: TObject);
begin
  EditChartTool(Self,Chart1.Tools[0]);
end;

procedure TSurfaceNearestForm.FormShow(Sender: TObject);
var P:TPoint;
begin
  inherited;

  // Move the mouse cursor to chart center
  P.X:=Chart1.ChartXCenter;
  P.Y:=Chart1.ChartYCenter;
  Mouse.CursorPos:=ClientToScreen(P);
end;

initialization
  RegisterClass(TSurfaceNearestForm);
end.
